guppylang/compiler/expr_compiler.py

```diff
@@ -533,6 +539,33 @@ def visit_BarrierExpr(self, node: BarrierExpr) -> Wire:
         self._update_inout_ports(node.args, iter(barrier_n), node.func_ty)
         return self._pack_returns([], NoneType())
 
+    def visit_StateResultExpr(self, node: StateResultExpr) -> Wire:
+        num_qubits = len(node.args) - 1
+        num_qubits_arg = ht.BoundedNatArg(num_qubits)
+        args = [ht.StringArg(node.tag), num_qubits_arg]
+        sig = ht.FunctionType(
+            [array_type(ht.Qubit, num_qubits_arg)],
+            [array_type(ht.Qubit, num_qubits_arg)],
+        )
+        op = ops.Custom(
+            op_name="StateResult", signature=sig, args=args, extension="tket2.debug"
+        )
+
+        qubits_in = [self.visit(e) for e in node.args[1:]]
+        # TODO: This should be caught during checking.
+        for q in qubits_in:
+            t = self.builder.hugr.port_type(q.out_port())
+            if t != ht.Qubit:
+                raise GuppyError(
+                    UnsupportedError(node, "Non-qubit inputs to state_result")
+                )
+        qubits_arr = self.builder.add_op(array_new(ht.Qubit, num_qubits), *qubits_in)
+        qubits_arr = self.builder.add_op(op, qubits_arr)
+        qubits_out = unpack_array(self.builder, qubits_arr)
```

Comment:
Not sure if using arrays in the hugr op is worth this overhead or there is a better way of representing the list of qubits in the extension

Reply:
I think this is fine for now. Would be a good idea to add an overload of `result_state` that takes an array anyways as a follow-up